graphql-protocol/src/main/java/com/linecorp/armeria/server/graphql/protocol/MultipartVariableMapper.java

```diff
@@ -0,0 +1,148 @@
+
+   private static final int INVALID_INT = -1;
+   private static final Splitter SPLITTER = Splitter.on(".");
+   private static final Mapper<Map<String, Object>> MAP_MAPPER = new Mapper<Map<String, Object>>() {
+      @Override
+      public Object set(Map<String, Object> location, String target, MultipartFile multipartFile) {
+         return location.put(target, multipartFile);
+      }
+
+      @Override
+      public Object recurse(Map<String, Object> location, String target) {
+         return location.get(target);
+      }
+   };
+
+   private static final Mapper<List<Object>> LIST_MAPPER = new Mapper<List<Object>>() {
+      @Override
+      public Object set(List<Object> location, String target, MultipartFile multipartFile) {
+         final int value = parseUnsignedInt(target);
+         if (value == INVALID_INT) {
+            return location;
+         }
+         if (value >= location.size()) {
+            return location;
+         }
+         return location.set(value, multipartFile);
+      }
+
+      @Override
+      public Object recurse(List<Object> location, String target) {
+         final int value = parseUnsignedInt(target);
+         if (value == INVALID_INT) {
+            return null;
+         }
+         if (value >= location.size()) {
+            return null;
+         }
+         return location.get(value);
+      }
+   };
+
+   static void mapVariable(String objectPath, Map<String, Object> variables, MultipartFile multipartFile) {
+      final List<String> segments = ImmutableList.copyOf(SPLITTER.split(objectPath));
+      if (segments.size() < 2) {
+         throw new IllegalArgumentException("Invalid object-path: " + objectPath);
+      }
+      if (!"variables".equals(segments.get(0))) {
+         throw new IllegalArgumentException("Can only map into variables: " + objectPath);
+      }
+
+      Object currentLocation = variables;
+      for (int i = 1; i < segments.size(); i++) {
+         final String segmentName = segments.get(i);
+         if (i == segments.size() - 1) {
+            if (currentLocation instanceof Map) {
+               if (MAP_MAPPER.set((Map<String, Object>) currentLocation, segmentName, multipartFile) != null) {
+                  throw new IllegalArgumentException("Expected null value when mapping: " + objectPath);
+               }
+            } else {
+               if (LIST_MAPPER.set((List<Object>) currentLocation, segmentName, multipartFile) != null) {
+                  throw new IllegalArgumentException("Expected null value when mapping: " + objectPath);
+               }
+            }
+         } else {
+            if (currentLocation instanceof Map) {
+               currentLocation = MAP_MAPPER.recurse((Map<String, Object>) currentLocation, segmentName);
+            } else {
+               currentLocation = LIST_MAPPER.recurse((List<Object>) currentLocation, segmentName);
+            }
+            if (currentLocation == null) {
```

Comment:
Would it make sense to move this `null` check outside this `else` block?

Reply:
I think the current position is correct, because I think it can return null in any case. what do you think? 🤔

Reply:
https://github.com/line/armeria/blob/0ebe09be1e30b8a7668a5875fcc9aee62ee6765c/graphql-protocol/src/main/java/com/linecorp/armeria/server/graphql/protocol/MultipartVariableMapper.java#L105-L114

I originally thought the following block could also result in `currentLocation == null`, but taking a closer look it seems otherwise 😅   I think it's fine to leave this in the current location 👍